### PR TITLE
Create stateless deploy playbook

### PR DESCRIPTION
**Pull request by haikuginger**
This playbook is designed for deploying a production OpenEdX instance using external databases and state, including Mongo, MySQL, ElasticSearch, Swift storage, and RabbitMQ.

This is based on the excellent work of @feanil in #3191, but aims to be a bit more explicit, and doesn't include automatic provisioning of the datastore instance(s).

**Dependencies**: Depends on #3349 for a Swift compatibility fix

**Testing instructions**:
1. Provision a server using this playbook. Be sure to include values for the following variables at a minimum:
   
   ``` yaml
   EDXAPP_MYSQL_DB_NAME: ''
   EDXAPP_MYSQL_USER: ''
   EDXAPP_MYSQL_PASSWORD: ''
   EDXAPP_MYSQL_HOST: ''
   EDXAPP_MYSQL_PORT: ''
   
   XQUEUE_MYSQL_DB_NAME: ''
   XQUEUE_MYSQL_USER: ''
   XQUEUE_MYSQL_PASSWORD: ''
   XQUEUE_MYSQL_HOST: ''
   XQUEUE_MYSQL_PORT: ''
   
   EDXAPP_MYSQL_CSMH_DB_NAME: ''
   EDXAPP_MYSQL_CSMH_USER: ''
   EDXAPP_MYSQL_CSMH_PASSWORD: ''
   EDXAPP_MYSQL_CSMH_HOST: ''
   EDXAPP_MYSQL_CSMH_PORT: ''
   
   EDX_NOTES_API_MYSQL_DB_NAME: ''
   EDX_NOTES_API_MYSQL_DB_USER: ''
   EDX_NOTES_API_MYSQL_DB_PASS: ''
   EDX_NOTES_API_MYSQL_HOST: ''
   
   COMMON_MYSQL_MIGRATE_USER: ''
   COMMON_MYSQL_MIGRATE_PASS: ''
   COMMON_MYSQL_READ_ONLY_USER: ''
   COMMON_MYSQL_READ_ONLY_PASS: ''
   COMMON_MYSQL_ADMIN_USER: ''
   COMMON_MYSQL_ADMIN_PASS: ''
   
   EDXAPP_MONGO_USER: ''
   EDXAPP_MONGO_PASSWORD: ''
   EDXAPP_MONGO_HOSTS: []
   EDXAPP_MONGO_PORT: ''
   EDXAPP_MONGO_DB_NAME: ''
   
   FORUM_MONGO_USER: ''
   FORUM_MONGO_PASSWORD: ''
   FORUM_MONGO_HOSTS: []
   FORUM_MONGO_PORT: ''
   FORUM_MONGO_DATABASE: ''
   
   EDXAPP_RABBIT_HOSTNAME: ''
   XQUEUE_RABBIT_HOSTNAME: ''
   
   EDXAPP_ELASTIC_SEARCH_CONFIG:
   - host: ''
     port: ''
   FORUM_ELASTICSEARCH_HOST: ''
   ```
2. Make changes on the server. Login, take a course, modify a course in studio.
3. Destroy the existing appserver, and provision a new one with the same settings.
4. Verify that the new server has the same data present as the previous server and is therefore using external database resources.

**Author notes and concerns**:
1. Note that if you're using OpenCraft IM to deploy, you'll get most of the above variables included for free. The exceptions are the Rabbit and ElasticSearch variables. For those, you'll need to add them to the `instance/templates/instance/ansible/vars.yml` template file or include them as custom arguments when spawning an instance. Additionally, you'll need to point OpenCraft IM at the `edx-stateless` playbook on this branch. Finally, be sure to use the OpenCraft IM "ephemeral databases" setting to generate external Mongo and MySQL database settings.

**Reviewers**
- [ ] @mtyaka
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fconfiguration&number=3370

**Comment by haikuginger** [comment]
@mtyaka, forgot to tag you in the PR at first, so this is just a ping.

File: playbooks/edx-stateless.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      nginx_sites:
+      - ecommerce
+      when: ENABLE_ECOMMERCE
+    - role: ecommerce
+      when: ENABLE_ECOMMERCE
+    - role: ecomworker
+      when: ENABLE_ECOMMERCE
+
+    # memcached
+    - memcache
```
Comment (mtyaka) [inline playbooks/edx-stateless.yml]: I think we should only install memcached when using local memcached, as in:
https://github.com/edx/configuration/blob/eb38e30a8d56484b593035d68b1807bdd13906d5/playbooks/edx_sandbox.yml#L38-L39

File: playbooks/edx-stateless.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  vars:
+    migrate_db: 'yes'
+    openid_workaround: True
+    EDXAPP_LMS_NGINX_PORT: '80'
+    ENABLE_ECOMMERCE: False  # Disable ecommerce by default
+  roles:
+
+    # Server setup
+    - swapfile
```
Comment (mtyaka) [inline playbooks/edx-stateless.yml:22]: Should we also add the `security` role?

Comment (haikuginger) [reply playbooks/edx-stateless.yml:22]: Yes we should. Good call.

File: playbooks/edx-stateless.yml (hunk 1 above)
Comment (haikuginger) [reply playbooks/edx-stateless.yml]: I like that pattern; sounds good to me. Memcache isn't truly stateless, but it also doesn't have real application state, so I'm fine with having it either way on this one.

**Comment by haikuginger** [comment]
@mtyaka, made the suggested changes.

**Comment by mtyaka** [comment]
Thanks @haikuginger!

:+1:
- [x] I tested this: Deployed a stateless instance using the new edx-stateless.yml playbook using a local installation of OpenCraft IM.
- [x] I read through the code
- [x] Includes inline documentation

**Comment by danielmcquillen** [comment]
@haikuginger I found when experimenting with small_deployments that I had to update the edxapp role so that EDXAPP_ELASTIC_SEARCH_CONFIG is written to both the lms _and_ cms (not just lms) env.json files, as otherwise when someone tries to index a course in Studio, the cms will throw an error since it will still expect elasticsearch to be at localhost. https://github.com/danielmcquillen/configuration/commit/30acc37cc7c86a72465651892581c847c207d6b4
Is that tweak applicable/helpful to this branch?

**Comment by haikuginger** [comment]
@danielmcquillen, well-spotted! Seems like something that's been lingering there a while, but that doesn't usually come up due to ES typically being hosted on the same instance as the appserver.

I've made the change and am testing a deploy to ensure it works.

**Comment by haikuginger** [comment]
I verified that deploy succeeded after making that change to the ES variable.

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @haikuginger! I've created [OSPR-1456](https://openedx.atlassian.net/browse/OSPR-1456) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

If you like, you can add yourself to the [AUTHORS](https://github.com/edx/configuration/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by gsong** [comment]
@haikuginger Do you think this will require documentation changes or it will be pretty transparent for most users?

**Comment by azizur** [comment]
I hope there will be some developer/ops documentations for those of us interested in deploying this playbook into our own environment.

**Comment by haikuginger** [comment]
@gsong: as far as documentation goes, it could swing either way. The playbook doesn't do anything fancy with its variables, so it should need minimal documentation; variables that work elsewhere should work here as well.

**Comment by haikuginger** [comment]
@gsong, any idea on when this might make it into review? Thanks!

**Comment by gsong** [comment]
@haikuginger It's scheduled for the current sprint, which ends on 2016-10-21.

File: playbooks/edx-stateless.yml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      - cms
+      - lms
+      - forum
+      - xqueue
+      nginx_default_sites:
+      - lms
+
+    # Swift storage compatibility
+    - role: openstack
+      when: EDXAPP_SETTINGS == 'openstack'
```
Comment (e0d) [inline playbooks/edx-stateless.yml]: I still don't like overloading the settings var as a feature flag.

File: playbooks/edx_sandbox.yml [hunk 4]
```diff
@@ -40,6 +40,8 @@
     - role: mongo
       when: "'localhost' in EDXAPP_MONGO_HOSTS"
     - { role: 'rabbitmq', rabbitmq_ip: '127.0.0.1' }
+    - role: openstack
+      when: EDXAPP_SETTINGS == 'openstack'
```
Comment (e0d) [inline playbooks/edx_sandbox.yml]: As above.

**Comment by e0d** [comment]
General comments:
- I'd like to give @feanil a chance to weigh in on this and he'll be back in a few weeks.
- Why wouldn't this include a stateful play?

**Comment by haikuginger** [comment]
@e0d, changes made. I've rebased from 6c49501 to include the latest changes in #3349 as a dependency for this PR.

> Why wouldn't this include a stateful play?

I'm not sure I understand this question. Could you clarify? Thanks!

**Comment by e0d** [comment]
@haikuginger can you rebase, not sure what's up with that license failure, but it's not happening on all PRs.  Also, @feanil  should be back early next week.

**Comment by haikuginger** [comment]
Rebased from 43bd0b0.

**Comment by e0d** [comment]
@feanil I think it makes sense for you to review this one now that you are back.

File: playbooks/roles/edxapp/defaults/main.yml [hunk 5]
```diff
@@ -949,6 +949,7 @@ generic_env_config:  &edxapp_generic_env
     limits: "{{ EDXAPP_CODE_JAIL_LIMITS }}"
     user: '{{ edxapp_sandbox_user }}'
   AFFILIATE_COOKIE_NAME: "{{ EDXAPP_AFFILIATE_COOKIE_NAME }}"
+  ELASTIC_SEARCH_CONFIG: "{{ EDXAPP_ELASTIC_SEARCH_CONFIG }}"
```
Comment (feanil) [inline playbooks/roles/edxapp/defaults/main.yml:958]: Why was elasticsearch moved to be available for both LMS and CMS?  Why does the CMS need it?

**Comment by feanil** [comment]
One question otherwise, generally looks good.

**Comment by haikuginger** [comment]
@feanil, I don't see a comment inline?

**Comment by feanil** [comment]
I asked: Why was elasticsearch moved to be available for both LMS and CMS? Why does the CMS need it?

**Comment by feanil** [comment]
On line 952 of edxapp/defaults/main.yml

**Comment by azizur** [comment]
@feanil did you happened to make a review comment and not submit it?

**Comment by danielmcquillen** [comment]
@feanil see my note farther up. In my setup (lms and cms on stateless, elastic search on datastores), the cms would throw an error when you try to re-index a course as by default it looks locally for elastic search. If you add edxapp role so that EDXAPP_ELASTIC_SEARCH_CONFIG gets set for cms, all seemed well. But I could have misunderstood something and therefore my suggestion to @haikuginger was incorrect.

**Comment by feanil** [comment]
@danielmcquillen I think I see what's going on, it looks like ES is needed if you are using either of the following features: `ENABLE_COURSEWARE_INDEX` or `ENABLE_LIBRARY_INDEX`

These are not on in our production environment and so we didn't notice this bug.  I'm good with moving the config so it's shared between the LMS and CMS as a default so that these features can work correctly.

**Review by feanil — APPROVED** [review]

**Comment by feanil** [comment]
@haikuginger I think this is good to go once you've rebased.

**Comment by haikuginger** [comment]
Rebased and squashed from 54db7cf.

**Comment by feanil** [comment]
:shipit:

**Pull request merged**